### ISSUE TITLE
discussion: redo 1200s https etc

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#50]
In case you were wondering about these redo section issues (there are more coming):
- the user.js has been *one* person's vision of how to structure it, order it, word it, where to put items, with some feedback in a wordpress forum, with over *two* years of adding items, usually at the end although I did try to leave gaps, and I did do some sort of revamps over the releases
- github affords us the chance to have really decent discussions (and arguments) in a proper collaborative environment, in order to whip this puppy into shape, and to overhaul all the inconsistencies, logic, wording, and more. The sky's the limit. Synergy and all that stuff.
- I want to revamp these sections (we don't need to do all of them) before I do a 52 release, that way, after the 52 release, the majority of changes will be minimal
- And after the revamps, Martin will do an article, and people who come and fork will then have less upheaval and commits etc.
- Basically, this is two years worth of everyone making my mess better :)

Section 1200 needs some love. I know @earthlng has some ideas for this section. I definitely want ciphers grouped together in say a 1250s. **This is all about the order, numbering, wording**. Don't want to get into discussions on the merits of turning things on and off or technical discussions on cipher suites and how curves are better than squares xD.

**Comment by earthlng** [comment]
IMO we covered all the changes in FF52 so we might as well do a release already.

**Comment by earthlng** [comment]
> the majority of changes will be minimal

maybe - maybe not. IMO we are not here to hold anyone's hand. It's very easy to compare versions on github

**Comment by earthlng** [comment]
You should focus on getting the readme finished and stuff like that, or we are unlikely to ever release a FF52 version before FF53 hits,

**Comment by Roman-Nopantski** [comment]
We have weeks and weeks and weeks (but I hear what you're saying - there's not that much I want to revamp). Until we get things into shape, it's actually more logical to delay the 52 release, as there are already massive changes from 51, so keep adding to them. Like I said I am not doing all sections. I'm also taking a break from wanting to write up the 3 wiki pages I need, and the readme (which will be simple and short - no instructions) - I'm enjoying these revamps.

I am thinking the release cycle is more likely to be at the end of a stable run, which is when it has had time to be tested and bugs ironed out and feedback received. Remember v10 and three or four items that caused issues, which led to adding warnings.

Cycle:
- A week before 52 comes out, we re-date the js and do a 52 release
- Then we start moving deprecated stuff down, adding 53 prefs and change the date and version in the header
- Repeat

I do have a plan, and a timetable, not just for this to be whipped and a 52 release, but also for martins article which will bring a lot of people I think, especially if we can get guys like Parker Lewis.

PS: Yeah, I've hit the close+comment by mistake a few times too :)

**Comment by earthlng** [comment]
> PS: Yeah, I've hit the close+comment by mistake a few times too :)

damn, you noticed that huh? :)

IMO a FF52 version should have all the FF52 pref-changes covered - it's just more logical

**Comment by Roman-Nopantski** [comment]
> IMO a FF52 version should have all the FF52 pref-changes covered - it's just more logical

I don't understand what you mean. Covered: yes we have gone thru the list. Covered as in tested: not yet (and in this small group of people it won't be, but future versions will have more feedback etc).

The "release" my friend, is an ARCHIVED version for legacy users. It should be given as much time as possible to be relevant and tested - and to get all the extras, and tweaks as possible that apply to it until the last second (or last week! xD). Meanwhile, the live version is ALREADY marked 52. It's not like this is a piece of compiled software and no one can use it until we do a release. Does that make more sense?

**Comment by Atavic** [comment]
```
/* 1213: disable 3DES (effective key size < 128)
* https://en.wikipedia.org/wiki/3des#Security
* http://en.citizendium.org/wiki/Meet-in-the-middle_attack
* http://www-archive.mozilla.org/projects/security/pki/nss/ssl/fips-ssl-ciphersuites.html ***/
```
URLs aren't 3DES specific, I'd move them and use the 2nd for MITM introduction (/* 1204:) while the 3rd could go as general intro (/*** 1200:) where there's
`https://www.securityartwork.es/2017/02/02/tls-client-fingerprinting-with-bro/`
that's really technical and maybe a let down as 1st link.
Drop wikipedia.

**Comment by Roman-Nopantski** [comment]
^^ I'll let earthlng comment on that

- Server side fingerprinting sucks. I think that we should match the default settings (but still keep entries, just comment them out as required) for ciphers. This would drastically lower the entropy. Sure, security comes first, and Mozilla can be a little slow (as can other browsers) to flip ciphers for fear of the internet breaking (no one wants to lose customers to the opposition). This means the header section can state that. I think the phrase "You can either strengthen your encryption/cipher suite and protocols.." is misleading and "Mozilla dragging their feet" is not fair. They have experts. This is not a indictment on PK's settings - his aim is to HARDEN and that's what he's doing by removing those 2 "least strong" (or whatever numbers they came in at) modern ciphers (they came in 7th and 9th of 9 or something). This would be a good balance - the default cipher suite is heavily scrutinized, and we also lower FP entropy.

- I will paste my outline for the re-ordering in the next post: the extra `// comment` lines should explain. I've already re-numbered them. I don't want to post 130lines, so I'll only do the numbered line and the actual prefs

**Comment by Roman-Nopantski** [comment]
```js
user_pref("ghacks_user.js.parrot", "1200 syntax error: the parrot's a stiff!");

// OCSP
/* 1201: enable OCSP Stapling (Online Certificate Status Protocol) ***/
user_pref("security.ssl.enable_ocsp_stapling", true);
/* 1202: query OCSP responder servers to confirm current validity of certificates (default=1) ***/
user_pref("security.OCSP.enabled", 1);
/* 1203: require certificate revocation check through OCSP protocol ***/
   // user_pref("security.OCSP.require", true);

// PINNING (is this even a subsectiony thing-what other secury prefs are coming up in 53?)
/* 1210: enforce strict pinning ***/
user_pref("security.cert_pinning.enforcement_level", 2);

// SSL
/* 1220: reject communication with servers using old SSL/TLS - vulnerable to a MiTM attack ***/
   // user_pref("security.ssl.require_safe_negotiation", true);
/* 1221: control TLS versions with min and max ***/
   // user_pref("security.tls.version.min", 2);
   // user_pref("security.tls.version.fallback-limit", 3);
   // user_pref("security.tls.version.max", 4); // 4 = allow up to and including TLS 1.3
/* 1222: disable SSL session tracking (FF36+) ***/
user_pref("security.ssl.disable_session_identifiers", true); // (hidden pref)

// HSTS
/* 1230: disable HSTS Priming (FF51+) ***/
user_pref("security.mixed_content.send_hsts_priming", false);
user_pref("security.mixed_content.use_hsts", false);
/* 1231: enforce HSTS preload list (default is true) ***/
user_pref("network.stricttransportsecurity.preloadlist", true);

// CERTS
/* 1240: disable Microsoft Family Safety cert (Windows 8.1) (FF50+) ***/
user_pref("security.family_safety.mode", 0);
/* 1241: disable intermediate certificate caching (fingerprinting attack vector) ***/
   // user_pref("security.nocertdb", true); // (hidden pref)

// CONTENT & UI
/* 1240: disable insecure active content on https pages - mixed content ***/
user_pref("security.mixed_content.block_active_content", true);
/* 1241: disable insecure passive content (such as images) on https pages - mixed context ***/
   // user_pref("security.mixed_content.block_display_content", true);
/* 1242: display warning (red padlock) for "broken security" ***/
user_pref("security.ssl.treat_unsafe_negotiation_as_broken", true);
/* 1243: control "Add Security Exception" dialog on SSL warnings ***/
user_pref("browser.ssl_override_behavior", 1);
/* 1244: display advanced information on Insecure Connection warning pages (thanks crssi) ***/
user_pref("browser.xul.error_pages.expert_bad_cert", true);

// CIPHERS
/* 1250: block rc4 fallback (default is now false as of at least FF45) ***/
user_pref("security.tls.unrestricted_rc4_fallback", false);
/* 1251: disable or limit SHA-1 ***/
user_pref("security.pki.sha1_enforcement_level", 1);
/* 1252: disable 3DES (effective key size < 128) ***/
user_pref("security.ssl3.rsa_des_ede3_sha", false);
/* 1253: disable 128 bits ***/
user_pref("security.ssl3.ecdhe_ecdsa_aes_128_sha", false);
user_pref("security.ssl3.ecdhe_rsa_aes_128_sha", false);
/* 1254: disable DHE (Diffie-Hellman Key Exchange) ***/
user_pref("security.ssl3.dhe_rsa_aes_128_sha", false);
user_pref("security.ssl3.dhe_rsa_aes_256_sha", false);
/* 1255: disable the remaining non-modern cipher suites as of FF52 ***/
   // user_pref("security.ssl3.rsa_aes_128_sha", false);
   // user_pref("security.ssl3.rsa_aes_256_sha", false);
```

**Comment by Roman-Nopantski** [comment]
items to consider if we use them

* security.tls.unrestricted_rc4_fallback - deprecated in 53
* security.pki.certificate_transparency.mode (FF52+)
* services.blocklist.pinning.* - do these go in 1200s? (FF53+)
* pref("security.webauth.webauthn", false); (FF53+)
pref("security.webauth.webauthn_enable_softtoken", false); pref("security.webauth.webauthn_enable_usbtoken", false);

**Comment by earthlng** [comment]
> The "release" my friend, is an ARCHIVED version for legacy users. ... Does that make more sense?
Yes, I understand now what your idea of a "release" is - but I disagree.
Since we aren't using v10, v11, etc anymore and instead basing it on the FF release, I think we should release a version right after FF stable is released and when we implemented all the changes from the diff and made sure that there are no more changes in the beta-to-stable-diff. So users can upgrade FF and get their update for the user.js with the same version.
I mean it's never really fully tested if we change something, in some cases. We try to pick out the prefs in a diff that we think need to be included and research what we think the value should be, but there can always be unpredictable behavior detected later on, or something we missed. Nobody can expect us to be perfect.
I've written up a draft for a "release changelog" and will post it in a separate issue - I think it may help to convince you of my idea.

**Comment by Atavic** [comment]
`services.blocklist.pinning` is about the blocklist.xml for malicious addons.

[// Blocklist via settings server (Kinto)](https://dxr.mozilla.org/mozilla-central/source/modules/libpref/init/all.js) at line 2224.

**Comment by Roman-Nopantski** [comment]
I'll read your release changelog and consider it. Obviously you get my points ( i hope). Don't confuse a github `release` with the user.js `version`. The github "release" is a one time snapshot for legacy users - they do not become legacy users until AFTER the next firefox stable comes out - so there is NO need to archive it until such time, which allows a full stable life cycle to test, get feedback, fix errors etc. The LIVE js is always stable, and the user.js is always in a state of flux. The only logical time to do the archiving is at end of life cycle.

**Comment by Roman-Nopantski** [comment]
OK, that's about the limit of my output. I am would like to leave in the naming of subsections, because this section is bit technical. I would do it like this

```js
/* OSCP (Online Certificate Status Protocol) ***/
/* 1201: enable OCSP Stapling ***/
```

If you want to merge subsections, change their names, switch the subsection order up, or radically redesign it, fire away.

Other than that, the only thing is deciding on reversing those 2 modern ciphers (and thus that then mean we are at all the same values as a nilla FF re ciphers)

**Comment by earthlng** [comment]
> The only logical time to do the archiving is at end of life cycle

I don't consider it an archiving but rather an updated version for the current FF release, for people that don't follow this closely and only get/want an updated user.js for each new FF stable release (at most)
Everyone who's using this, should always use the latest stable FF release (if they're not using Beta, Nightly, ESR) - so it doesn't make sense to me that the release should be using the same version as the current FF stable, but be used by people on older versions. Or are you saying that as soon as FF53 is released, people should start using our v52. (?) That's why I said it's not logical to me.

> legacy users

there should never be any legacy users

**Comment by earthlng** [comment]
```
/* 1213: disable 3DES (effective key size < 128)
* https://en.wikipedia.org/wiki/3des#Security
* http://en.citizendium.org/wiki/Meet-in-the-middle_attack
* http://www-archive.mozilla.org/projects/security/pki/nss/ssl/fips-ssl-ciphersuites.html ***
```
> URLs aren't 3DES specific, I'd move them and use the 2nd for MITM introduction (/* 1204:) while the 3rd could go as general intro (/*** 1200:) where now there is
> https://www.securityartwork.es/2017/02/02/tls-client-fingerprinting-with-bro/
> that's really technical and maybe a let down as 1st link.
> 
> Drop wikipedia.

IMO the wiki-link is the only link that illustrates why we disable 3des, with only a few sentences.
The 3rd link is IMO way more confusing and hard to understand, and we can drop that one.
And the second is `Meet-in-the-middle_attack` **not** `Man-in-the-middle`. That link should definitely stay, but we'll see where it fits best.
I agree the securityartwork.es article is quite technical once it gets to the "Enter Bro" part. But you can stop reading at that point anyway. But your point is taken - I'll create a wiki with quotes and the images from that article and we can link to our wiki instead. (once done, I'll ask the author of the article if he minds that we quote him and use his images for our wiki)

Thank you @Atavic

**Comment by earthlng** [comment]
> I'll create a wiki with quotes

I think we can just add a note to the link that `the article is quite technical but the first part is easy to understand and you can stop reading at "Enter Bro" `.
It's only the first 10 sentences or so, that matter and illustrate the problem.

**Comment by earthlng** [comment]
> Don't want to get into discussions on the merits of turning things on and off 

> I think that we should match the default settings (but still keep entries, just comment them out as required) for ciphers.

classic Pants ;)

**Comment by earthlng** [comment]
> outline for the re-ordering

I definitely agree with grouping the ciphers together. But for the rest - I'm also not exactly an expert - Idk which ones make sense to group together. One group I could see is mixed-content.
But it makes it more complicated when we never really know where something fits best. So I say we keep the rest un-grouped.

> deciding on reversing those 2 modern ciphers (and thus that then mean we are at all the same values as a nilla FF re ciphers)

1261+1262+1263 are all still enabled by default, so those would all need to be commented out to achieve `same values as a nilla FF re ciphers`

**Comment by Roman-Nopantski** [comment]
> Don't want to get into discussions on the merits of turning things on and off

"don't want" and "will not" are two different things :) I only brought the ciphers up because you have (rightly) been harping on about those two modern ciphers being disabled, so we might as well do em in this revamp. And I see now that you checked what had to be done to achieve same as FF ciphers, which looks like the lot of em. Do we flick them all to commented out (except SHA-1, the default is 3 which is local certs, so no leakage there), and move the CA fingerprinting info down to the sub-header. I'm actually happy with Mozilla's defaults. So basically comment out 5 ciphers - essentially leaving a section for an uber-hardened js.

As for the subsections, yeah - some expert will come along and say "what idjuts, don't they know how a time machine works". Everything will somehow be interconnected and ours is just one way of looking at it. I think a couple of them could be combined, but at least if I leave gaping holes in the numbers that will help in future. PS: mixed content is already grouped

**Comment by Roman-Nopantski** [comment]
will go with, unless objections/corrections. Should we change the order, like make SSL first. Should we change the 1200 title to match the order?

```js
/* OCSP (Online Certificate Status Protocol) ***/
/* SSL (Secure Sockets Layer) / TLS (Transport Layer Security) ***/
/* HSTS (HTTP Strict Transport Security) ***/
/* CERTS / PINNING / HPKP (HTTP Public Key Pinning) ***/
/* CONTENT & UI ***/
/* CIPHERS ***/
```

**Comment by Roman-Nopantski** [comment]
Just a quick note (I really have to get back to my paid work for a while) ... 

- this is just the sort of page I hope @fmarier (since he's awake right now) jumps in and corrects us if we fubar
- We should check the cipher suite TBB uses
- thinking along lines of lite-medium-hardened for future eg like if medium comments out those 0850+'s, this would be another one, we comment out the ciphers - i.e [H] only: I trust Mozilla's choices for a medium setting and it would actually reduce the FP

**Comment by earthlng** [comment]
> Do we flick them all to commented out (except SHA-1, the default is 3 which is local certs, so no leakage there), and move the CA fingerprinting info down to the sub-header

sure, why not. The comic with the wrench comes to mind :)

> I'm actually happy with Mozilla's defaults.

I'm not exactly happy with the defaults but I understand why mozilla does it.

> PS: mixed content is already grouped

unless you changed something since your proposal - no.
all the `security.mixed_content.*` should be grouped together because 1220 is first and foremost a mixed-content "improvement" (hence the name branch)
That would leave 1252+1253+1254 as `UI` prefs - I'd be happy with that.

```
/* SSL (Secure Sockets Layer) / TLS (Transport Layer Security) ***/
/* OCSP (Online Certificate Status Protocol) ***/
/* CERTS / PINNING ***/
/* HSTS (HTTP Strict Transport Security) / HPKP (HTTP Public Key Pinning) ***/
/* MIXED CONTENT ***/
/* CIPHERS ***/
/* UI (User Interface) ***/
```
HSTS and HPKP (no pref for HPKP, correct?) could/should go together because they're also stored in the same file at the moment. (it would only have 1 pref in that section atm, right? - but regardless, that grouping makes sense IMO)
UI last because that's the least important

**Comment by Roman-Nopantski** [comment]
my numbers are different now .. there are four `security.mixed_content.` prefs

- `security.mixed_content.block_active_content` & `security.mixed_content.block_display_content` are grouped together - under content, that's what I thought you were talking about.
- `security.mixed_content.send_hsts_priming` and `security.mixed_content.use_hsts` are grouped under HSTS
- what the heck am i missing that u are trying to say

no pref for hpkp - not that i can see, i stuck it with pinning because it was pinning. no big deal with your order as ocsp (certs related) leads to certs/pinning which leads to hkpk which is pinning but also grouped with hsts .. idk .. everything is HSTS .. like I said, everything is interconnected and right now the time lords are laughing at us. Any way we slice and dice it will be fine.

**Comment by Roman-Nopantski** [comment]
not final, just a complete new draft for referal

```js
/*** 1200: HTTPS ( SSL/TLS / HSTS / OCSP / CERTS / HPKP / ENCRYPTION )
     Note that your cipher and other settings can be used server side as a fingerprint attack vector:
     see https://www.securityartwork.es/2017/02/02/tls-client-fingerprinting-with-bro/
     You can either strengthen your encryption/cipher suite and protocols (security) or keep them
     at default and let Mozilla handle them (dragging their feet for fear of breaking legacy sites)
 ***/
user_pref("ghacks_user.js.parrot", "1200 syntax error: the parrot's a stiff!");

/* SSL (Secure Sockets Layer) / TLS (Transport Layer Security) ***/
/* 1201: reject communication with servers using old SSL/TLS - vulnerable to a MiTM attack ***/
   // user_pref("security.ssl.require_safe_negotiation", true);
/* 1202: control TLS versions with min and max ***/
   // user_pref("security.tls.version.min", 2);
   // user_pref("security.tls.version.fallback-limit", 3);
   // user_pref("security.tls.version.max", 4); // 4 = allow up to and including TLS 1.3
/* 1203: disable SSL session tracking (FF36+) ***/
user_pref("security.ssl.disable_session_identifiers", true); // (hidden pref)

/* OCSP (Online Certificate Status Protocol) ***/
/* 1210: enable OCSP Stapling ***/
user_pref("security.ssl.enable_ocsp_stapling", true);
/* 1211: query OCSP responder servers to confirm current validity of certificates (default=1) ***/
user_pref("security.OCSP.enabled", 1);
/* 1212: require certificate revocation check through OCSP protocol ***/
   // user_pref("security.OCSP.require", true);

/* CERTS / PINNING ***/
/* 1220: disable Microsoft Family Safety cert (Windows 8.1) (FF50+) ***/
user_pref("security.family_safety.mode", 0);
/* 1221: disable intermediate certificate caching (fingerprinting attack vector) ***/
   // user_pref("security.nocertdb", true); // (hidden pref)
/* 1225: enforce strict pinning ***/
user_pref("security.cert_pinning.enforcement_level", 2);

/* HSTS (HTTP Strict Transport Security) / HPKP (HTTP Public Key Pinning) ***/
/* 1230: disable HSTS Priming (FF51+) ***/
user_pref("security.mixed_content.send_hsts_priming", false);
user_pref("security.mixed_content.use_hsts", false);
/* 1231: enforce HSTS preload list (default is true) ***/
user_pref("network.stricttransportsecurity.preloadlist", true);

/* MIXED CONTENT ***/
/* 1240: disable insecure active content on https pages - mixed content ***/
user_pref("security.mixed_content.block_active_content", true);
/* 1241: disable insecure passive content (such as images) on https pages - mixed context ***/
   // user_pref("security.mixed_content.block_display_content", true);

/* CIPHERS ***/
/* 1260: disable or limit SHA-1
 * 0 = all SHA1 certs are allowed
 * 1 = all SHA1 certs are blocked (including perfectly valid ones from 2015 and earlier)
 * 2 = deprecated option that now maps to 1
 * 3 = only allowed for locally-added roots (e.g. anti-virus)
 * 4 = only allowed for locally-added roots or for certs in 2015 and earlier
 * [WARNING] when disabled, some man-in-the-middle devices (eg security scanners and antivirus
 * products, are failing to connect to HTTPS sites. SHA-1 will eventually become obsolete.
 * https://blog.mozilla.org/security/2016/10/18/phasing-out-sha-1-on-the-public-web/
 * https://github.com/pyllyukko/user.js/issues/194#issuecomment-256509998 ***/
user_pref("security.pki.sha1_enforcement_level", 1);
/* 1261: disable 3DES (effective key size < 128)
 * https://en.wikipedia.org/wiki/3des#Security
 * http://en.citizendium.org/wiki/Meet-in-the-middle_attack
 * http://www-archive.mozilla.org/projects/security/pki/nss/ssl/fips-ssl-ciphersuites.html ***/
user_pref("security.ssl3.rsa_des_ede3_sha", false);
/* 1262: disable 128 bits ***/
user_pref("security.ssl3.ecdhe_ecdsa_aes_128_sha", false);
user_pref("security.ssl3.ecdhe_rsa_aes_128_sha", false);
/* 1263: disable DHE (Diffie-Hellman Key Exchange)
 * https://www.eff.org/deeplinks/2015/10/how-to-protect-yourself-from-nsa-attacks-1024-bit-DH
 * [WARNING] may break obscure sites, but not major sites, which should support ECDH over DHE ***/
user_pref("security.ssl3.dhe_rsa_aes_128_sha", false);
user_pref("security.ssl3.dhe_rsa_aes_256_sha", false);
/* 1264: disable the remaining non-modern cipher suites as of FF52
 * [NOTE] commented out because it still breaks too many sites ***/
   // user_pref("security.ssl3.rsa_aes_128_sha", false);
   // user_pref("security.ssl3.rsa_aes_256_sha", false);
/* 1265: block rc4 fallback (will be deprecated in 53) ***/
user_pref("security.tls.unrestricted_rc4_fallback", false);

/* UI (User Interface) ***/
/* 1270: display warning (red padlock) for "broken security" ***/
user_pref("security.ssl.treat_unsafe_negotiation_as_broken", true);
/* 1271: control "Add Security Exception" dialog on SSL warnings ***/
user_pref("browser.ssl_override_behavior", 1);
/* 1272: display advanced information on Insecure Connection warning pages ***/
user_pref("browser.xul.error_pages.expert_bad_cert", true);
```

**Comment by earthlng** [comment]
> `security.mixed_content.send_hsts_priming` and `security.mixed_content.use_hsts` are grouped under HSTS

yes thats the problem. they are a mixed-content "improvement". They don't fit under HSTS, because when you don't block mixed content those 2 are irrelevant, see what I mean?

> But it makes it more complicated when we never really know where something fits best. So I say we keep the rest un-grouped.

I don't see any problems with the following groups but the rest should maybe best remain un-sub-grouped, because

> everything is interconnected

```
/* MIXED CONTENT ***/ -> for the 4 mixed-content prefs
/* CIPHERS ***/
/* UI (User Interface) ***/ -> 1252+1253+1254
```

I think it would be best if we do these kinds of discussions in a pull request. We can do 100 commits in there if necessary and as soon as we agreed on something merge that one commit into the PR, etc.
For example we both agree on the Ciphers sub-group.
When everything is done, just merge the agreed upon PR into the master and finito.

**Comment by Atavic** [comment]
>no pref for HPKP, correct?

/*** 2698: FIRST PARTY ISOLATION (FPI) ***/ has only a resolved bugzilla reference.

**Comment by Roman-Nopantski** [comment]
I hate pull requests

**Comment by Roman-Nopantski** [comment]
`send_hsts_priming` and `use_hsts` are HSTS .. but the first part is `mixed_content` ... either way we can't win. I will make one last change - recheck the draft above in 2 minutes.

**Comment by earthlng** [comment]
I could you piss you off sooo much by quoting what you recently said in another thread :)

**Comment by fmarier** [comment]
> /* 1242: disable HSTS Priming (FF51+) ***/
> user_pref("security.mixed_content.send_hsts_priming", false);
> user_pref("security.mixed_content.use_hsts", false);

What's the benefit in disabling this?

With `use_hsts` off, the browser enforces mixed content restrictions even though they wouldn't actually lead to mixed content (given the site is already known to be HSTS-enabled). With `use_hsts` on, the restrictions are removed in some of the cases where the browser already knows it's not going to be mixed content.

Priming is an extra check which will increase the number of sites that are discovered to be hsts-enabled. This seems like a clear security win, and is the reason why we are experimenting with this idea.

**Comment by earthlng** [comment]
so as for the changes that @Atavic suggested. Do we want to include something like this for the first link:
`the article is quite technical but the 1st part is easy to understand and you can stop reading at "Enter Bro"`

We can remove the 3rd link in 1261, because it's very old, quite a mess, hard to understand, doesn't really add anything and half the page is (OBSOLETE). The 1st and 2nd link should stay under 3des however.

you also didn't want to blame mozilla for dragging their feet anymore but it's still there.

> and move the CA fingerprinting info down to the sub-header

no. it also applies to other things afaik. fe. disabling session-tickets. Maybe @fmarier knows more about this.

**Comment by earthlng** [comment]
> What's the benefit in disabling HSTS Priming?

> With `use_hsts` off, the browser enforces mixed content restrictions even though they wouldn't actually lead to mixed content

With all due respect but if that's really the case then that's a terrible implementation.
If a new feature is turned off it should have exactly zero effects IMHO.

[This](http://www.ghacks.net/2015/08/18/a-comprehensive-list-of-firefox-privacy-and-security-settings/#comment-4116036) is why we decided to turn HSTS Priming off.

**Comment by Roman-Nopantski** [comment]
@earthlng - the paste was just for the ordering sorry (got lazy) - we have yet to talk about what to do with ciphers being active/inactive. And the stuff in the header still needs looking at (yes, it can stay in the header already spotted that) - and yes, butt heads, but we're going to go around in circles.

**Comment by Atavic** [comment]
```
/*** 1200: HTTPS ( SSL/TLS / OCSP / CERTS / HSTS / HPKP / ENCRYPTION )
     Note that cipher suites can be used server side for fingerprinting, see:
     https://www.owasp.org/index.php/Testing_for_weak_Cryptography
     You can drop weak cryptographic primitives (security) or keep them
     at default (Mozilla keeps some old cipher suites until they are expired)
 ***/
```
...kept it short, but

https://www.owasp.org/index.php/Testing_for_weak_Cryptography

should really be:

https://www.owasp.org/index.php/Testing_for_Weak_SSL/TLS_Ciphers,_Insufficient_Transport_Layer_Protection_(OTG-CRYPST-001)

and 

(Mozilla keeps some old cipher suites until they are expired)

should really be:

(Mozilla keeps some old cipher suites until the expiration date of the CAs containing them is reached)

**Comment by crssi** [comment]
@earthlng when you mentioned OBSOLETE... wouldn't it be maybe nicer if obsoleted content would be archived into another file, like obsoleted.js to clenaup the current one?

**Comment by earthlng** [comment]
@Atavic, wait, you thought the Spain (.es) link was too technical but now you link us to this 10x longer article with nmap outputs and shit like that? How is that not waaaaay worse?
Also IMO there aren't any **weak** ciphers in FF atm. Yes, some are better than others, but they aren't weak. It'll take me a while to read thru that link, but if you have already done so please point me to where they say that one or more of the current ciphers in FF are considered as weak

**Comment by earthlng** [comment]
@crssi, look at [this](http://www-archive.mozilla.org/projects/security/pki/nss/ssl/fips-ssl-ciphersuites.html). That's what I was referring to.

**Comment by crssi** [comment]
@earthlng somehow I knew I sad something wrong, but didn't realized at the time...lol

**Comment by earthlng** [comment]
@crssi, please don't be mad at me but I'm deleting your comments and my reply, to keep this as clean and relevant as possible. I'll also delete this very comment again as soon as you give me a thumbs-up to let me know that you're not mad at me :)

**Comment by fmarier** [comment]
> With all due respect but if that's really the case then that's a terrible implementation. If a new feature is turned off it should have exactly zero effects IMHO.

What I was describing is not a new behavior. Whether you use an old browser or a new version of Firefox with that setting disabled, it's the same thing: the mixed content blocker kicks in to block a load that would be done over HTTPS.

> [This](http://www.ghacks.net/2015/08/18/a-comprehensive-list-of-firefox-privacy-and-security-settings/#comment-4116036) is why we decided to turn HSTS Priming off.

The goal of the mixed content blocker (MCB) is to block HTTP requests from HTTPS pages. It's not there to reduce data usage or block trackers. That's the role of ublock, adblockplus, request policy, tracking protection, etc.

So what `use_hsts` does is reduce the number of false positives. It reduces the amount of unnecessary breakage that the MCB causes while still ensuring that no HTTP connections are made from HTTPS pages.

The pre-flight request it adds (to a host it would already need to connect to if it displayed the content) is over HTTPS, so again, it's not violating what it's supposed to do. It's merely trying to make the MCB more useful, which could help us make it more aggressive in the future (for example, blocking passive mixed content by default).

**Comment by Roman-Nopantski** [comment]
you do realize that they are both false by default, right? If they get flipped in a future release, we can revisit it (and probably turn them on?)

user_pref("security.mixed_content.send_hsts_priming", false);
user_pref("security.mixed_content.use_hsts", false);

**Comment by fmarier** [comment]
> you do realize that they are both false by default, right?

I primarily use Firefox Nightly where they are both enabled by default :D

**Comment by Roman-Nopantski** [comment]
^^ @fmarier : totally get that :) We pick up on all changes in every release (including minor), and when we see them change, we will probably enforce them (rather than a commented out disable)

--
> Also IMO there are no weak ciphers in FF atm

@earthlng  That's what I'm saying. I have no problems with this js (as compared to our hardened one) being all ciphers at default (except sha-1 which is default local anyway). Also, all the SSL is at default. I can't see what else could be used to FP (did that test use mixed active/mixed passive? they could add in session ticket checking).

So it's an all or nothing in my opinion - lower entropy against server side FPing (which ANY site could implement), or harden up a little on some ciphers which pose an extremely small risk IMO

Am interested in what Francois thinks (and uses)

**Comment by earthlng** [comment]
> the mixed content blocker kicks in to block a load that would be done over HTTPS.

What? Why does MCB block a load over HTTPS? The goal of MCB is to block HTTP requests, not HTTPS ones, as you said yourself in your next sentence. Please elaborate. Or just a typo?
If its a typo, I still don't understand when you said

> the browser enforces mixed content restrictions even though they wouldn't actually lead to mixed content

so please elaborate either way.

---

> It's not there to reduce data usage or block trackers

I never said that. I'd appreciate if you could quote from my assessment and tell me where I'm wrong.

> it's not violating what it's supposed to do

I disagree. If I block *all* mixed-content, why would I want additional requests to detect if a HTTP-resource is also available over HTTPS? for convenience only, right? To perhaps unblock a picture here or there that would otherwise be blocked. And that's exactly what I said in my assessment.

Ok, maybe it's not **violating** what it's supposed to do, but it works around what it **really** is supposed to do. "It changes the order" to quote from the bugzilla and the patch author.
Yes, I know the requests are still done over HTTPS, but still, how is that not a convenience feature.
I see what you mean, that in a very broad sense it can be considered a security-feature because it helps you move more people to block mixed-content. But until you enable passive mixed-content blocking by default, HSTS priming works around MCB and is just not a security feature.

And IMO it'll take years before mozilla will dare to enable passive mixed-content blocking by default, but I also like surprises, so... keep at it ;)

> It's merely trying to make the MCB more useful

Exactly. It's a convenience feature.

> Mixed-content blocking may prevent some sites from moving from HTTP to HTTPS. In order to help sites opportunistically move to HTTPS, we introduce the concept of HSTS Priming.

I just really hope you keep the prefs available to disable HSTS priming.

**Comment by earthlng** [comment]
we originally added it with the following note, and if anything is incorrect about it in your opinion, please elaborate

```
// RISKS: formerly blocked mixed-content may load, may cause noticeable delays eg requests
// time out, requests may not be handled well by servers, possible fingerprinting
```

**Comment by Atavic** [comment]
@earthlng The long OWASP URL explains a lot in a better way than the page about Bro.
And 'weak' was just a general term, as with new CPUs old cyphers are considered weak.
Call it 'problematic' if you want.

**Comment by fmarier** [comment]
>> the mixed content blocker kicks in to block a load that would be done over HTTPS.
>
> What? Why does MCB block a load over HTTPS? The goal of MCB is to block HTTP requests, not HTTPS ones, as you said yourself in your next sentence. Please elaborate. Or just a typo?
If its a typo, I still don't understand when you said

Take this example:

* `https://example.com` is an HTML page which loads `http://example.net/example.js`

That's active mixed content and it's blocked by the MCB because if you were to disable MCB, the browser would do an HTTP load from an HTTPS page and that's what the MCB is supposed to block.

Now, let's look at this different example:

* `https://example.com` is an HTML page which loads `http://example.net/example.js`
* `http://example.net/*` actually redirects to `https://example.net/*`
* `https://example.net/*` sets an HSTS header

We still have something that looks like active mixed content and so the MCB kicks in and blocks the script load. However, if you were to turn off MCB, what you would get is `https://example.com` loading `http**s**://example.net/example.js` directly, not `http://example.net/example.js`.

Therefore, MCB blocks an HTTPS to HTTPS request just because it runs **before** HSTS. The idea of the `use_hsts = true` behavior is to check for HSTS before MCB to avoid unnecessary blocking. In a way, it's kind of a bug in the MCB specification and we are hoping to fix it there.

**Comment by earthlng** [comment]
Ok, thanks, I understand what you mean now.
But with MCB enabled it would only work that way if I had previously visited `https://example.net/` so that it could have set an entry in my HSTS cache, correct?
If so, how does that work when the HSTS cache is isolated thanks to the Tor Uplift project?

> In a way, it's kind of a bug in the MCB specification and we are hoping to fix it there.

I wouldn't say it's a bug. MCB works exactly as it's supposed to work IMHO. ie see http -> discard

One could say you're encouraging lazy web-admins to never fix their sites, if you make things work when they really shouldn't.

And while I totally agree that HEAD requests are way better (because faster), they are also fairly rare and

> stand out in most servers logfiles like a black guy at a KKK meeting

as I so eloquently put it ;) so it adds a few more bits for fingerprinting. Add the 24h cache to that, and I'm sure the TBB guys are not amused.

**Comment by fmarier** [comment]
> But with MCB enabled it would only work that way if I had previously visited `https://example.net/` so that it could have set an entry in my HSTS cache, correct?

In the absence of priming, `use_hsts` relies on:
- a previous a visit to the `https://example.net/` homepage (like you said),
- a previous image load (not blocked by MCB since it's passive) from `http://example.net`, which got redirected by the server to `https://example.net`,
- or the `example.net` domain being in the HSTS preload list.

> If so, how does that work when the HSTS cache is isolated thanks to the Tor Uplift project?

Non-preloaded URLs are much less likely to be in the HSTS cache in that case because the HSTS cache is not shared across domains.

**Comment by earthlng** [comment]
Thanks, final questions if you don't mind ...

Do you understand now why we disable it, and do my reasons make even the slightest of sense to you?

**Comment by earthlng** [comment]
@fmarier one way to make MCB better and something I wouldn't mind seeing "fixed" directly in the MCB itself, would be to upgrade HTTP to HTTPS for resources on domains that already have a HSTS entry in the cache or the preload list. Perhaps **only** if the domain has the `includeSubDomains` directive which should guarantee that the resource is available over HTTPS. That way no additional requests would be necessary and MCB could redirect internally to the HTTPS resource. I just really don't like the additional "priming" requests. Since the `includeSubDomains` is mandatory for domains on the preload list, maybe you could start with that.

**Comment by Thorin-Oakenpants** [comment]
final FULL patch for nits

```js
/*** 1200: HTTPS ( SSL/TLS / OCSP / CERTS / HSTS / HPKP / ENCRYPTION )
   Note that your cipher and other settings can be used server side as a fingerprint attack
   vector, see [1]. This article is quite technical but the first part is easy to understand
   and you can stop reading when you reach the second section titled "Enter Bro".

   Option 1: Use our settings to tighten up encryption options. It *is* a fingerprinting attack
             vector, and we certainly do want to reduce any attack surface, but this is not how
             you *DEFEAT* fingerprinting - to do that you need large numbers to buy into the same
             enforced browser-wide settings (such as TBB), and/or you use OpSec.
   Option 2: Use Firefox defaults for the 1260's items (item 1260 default for SHA-1, is local only
             anyway). There is nothing *weak* about Firefox's defaults, but Mozilla (and other
             browsers) will always lag for fear of breakage and upset end-users

   [1] https://www.securityartwork.es/2017/02/02/tls-client-fingerprinting-with-bro/
 ***/
user_pref("ghacks_user.js.parrot", "1200 syntax error: the parrot's a stiff!");
/* SSL (Secure Sockets Layer) / TLS (Transport Layer Security) ***/
/* 1201: reject communication with servers using old SSL/TLS - vulnerable to a MiTM attack
 * [WARNING] tested Feb 2017 - still breaks too many sites
 * [1] https://wiki.mozilla.org/Security:Renegotiation ***/
   // user_pref("security.ssl.require_safe_negotiation", true);
/* 1202: control TLS versions with min and max
 * 1=min version of TLS 1.0, 2-min version of TLS 1.1, 3=min version of TLS 1.2 etc
 * [WARNING] FF/chrome currently allow TLS 1.0 by default, so this is your call.
 * [1] http://kb.mozillazine.org/Security.tls.version.*
 * [2] https://www.ssl.com/how-to/turn-off-ssl-3-0-and-tls-1-0-in-your-browser/ ***/
   // user_pref("security.tls.version.min", 2);
   // user_pref("security.tls.version.fallback-limit", 3);
   // user_pref("security.tls.version.max", 4); // 4 = allow up to and including TLS 1.3
/* 1203: disable SSL session tracking (FF36+)
 * SSL Session IDs speed up HTTPS connections (no need to renegotiate) and last for 48hrs.
 * Since the ID is unique, web servers can (and do) use it for tracking. If set to true,
 * this disables sending SSL Session IDs and TLS Session Tickets to prevent session tracking
 * [1] https://tools.ietf.org/html/rfc5077
 * [2] https://bugzilla.mozilla.org/show_bug.cgi?id=967977 ***/
user_pref("security.ssl.disable_session_identifiers", true); // (hidden pref)
/* OCSP (Online Certificate Status Protocol) ***/
/* 1210: enable OCSP Stapling
 * [1] https://blog.mozilla.org/security/2013/07/29/ocsp-stapling-in-firefox/ ***/
user_pref("security.ssl.enable_ocsp_stapling", true);
/* 1211: query OCSP responder servers to confirm current validity of certificates
 * 0=disable, 1=validate only certificates that specify an OCSP service URL (default)
 * 2=enable and use values in security.OCSP.URL and security.OCSP.signing ***/
user_pref("security.OCSP.enabled", 1);
/* 1212: require certificate revocation check through OCSP protocol
 * This leaks information about the sites you visit to the CA (cert authority)
 * It's a trade-off between security (checking) and privacy (leaking info to the CA)
 * [WARNING] Since FF44 the default is false. If set to true, this may/will cause some
 * site breakage. Some users have previously mentioned issues with youtube, microsoft etc ***/
   // user_pref("security.OCSP.require", true);
/* CERTS / HSTS (HTTP Strict Transport Security) / HPKP (HTTP Public Key Pinning) ***/
/* 1220: disable Microsoft Family Safety cert (Windows 8.1) (FF50+)
 * 0 = disable detecting Family Safety mode and importing the root
 * 1 = only attempt to detect Family Safety mode (don't import the root)
 * 2 = detect Family Safety mode and import the root ***/
user_pref("security.family_safety.mode", 0);
/* 1221: disable intermediate certificate caching (fingerprinting attack vector)
 * [NOTE] This may be better handled under FPI (ticket 1323644, part of Tor Uplift)
 * [WARNING] This affects login/cert/key dbs. The effect is all credentials are session-only.
 * Saved logins and passwords are not available. Reset the pref and restart to return them.
 * [1] https://bugzilla.mozilla.org/show_bug.cgi?id=1334485 - related bug
 * [2] https://bugzilla.mozilla.org/show_bug.cgi?id=1216882 - related bug (see comment 9) ***/
   // user_pref("security.nocertdb", true); // (hidden pref)
/* 1222: enforce strict pinning
 * PKP (Public Key Pinning) 0=disabled 1=allow user MiTM (such as your antivirus), 2=strict
 * [WARNING] If you rely on an AV (antivirus) to protect your web browsing
 * by inspecting ALL your web traffic, then leave at current default=1
 * [1] https://trac.torproject.org/projects/tor/ticket/16206 ***/
user_pref("security.cert_pinning.enforcement_level", 2);
/* 1223: enforce HSTS preload list (default is true)
 * The list is compiled into Firefox and is used to always use HTTPS for the domains on that list
 * [1] https://blog.mozilla.org/security/2012/11/01/preloading-hsts/
 * [2] https://wiki.mozilla.org/Privacy/Features/HSTS_Preload_List ***/
user_pref("network.stricttransportsecurity.preloadlist", true);
/* MIXED CONTENT ***/
/* 1240: disable insecure active content on https pages - mixed content ***/
user_pref("security.mixed_content.block_active_content", true);
/* 1241: disable insecure passive content (such as images) on https pages - mixed context
 * current default=false, leave it this way as too many sites break visually ***/
   // user_pref("security.mixed_content.block_display_content", true);
/* 1242: disable HSTS Priming (FF51+)
 * We disable it because formerly blocked mixed-content may load, may cause noticeable delays
 * eg requests time out, requests may not be handled well by servers, possible fingerprinting
 * [1] https://bugzilla.mozilla.org/show_bug.cgi?id=1246540#c145 ***/
   // user_pref("security.mixed_content.send_hsts_priming", false);
   // user_pref("security.mixed_content.use_hsts", false);
/* CIPHERS [see the section 1200 intro] ***/
/* 1260: disable or limit SHA-1
 * 0 = all SHA1 certs are allowed
 * 1 = all SHA1 certs are blocked (including perfectly valid ones from 2015 and earlier)
 * 2 = deprecated option that now maps to 1
 * 3 = only allowed for locally-added roots (e.g. anti-virus)
 * 4 = only allowed for locally-added roots or for certs in 2015 and earlier
 * [WARNING] when disabled, some man-in-the-middle devices (eg security scanners and
 *  antivirus products, may fail to connect to HTTPS sites. SHA-1 is *almost* obsolete.
 * [1] https://blog.mozilla.org/security/2016/10/18/phasing-out-sha-1-on-the-public-web/ ***/
user_pref("security.pki.sha1_enforcement_level", 1);
/* 1261: disable 3DES (effective key size < 128)
 * [1] https://en.wikipedia.org/wiki/3des#Security
 * [2] http://en.citizendium.org/wiki/Meet-in-the-middle_attack
 * [3] http://www-archive.mozilla.org/projects/security/pki/nss/ssl/fips-ssl-ciphersuites.html ***/
user_pref("security.ssl3.rsa_des_ede3_sha", false);
/* 1262: disable 128 bits ***/
user_pref("security.ssl3.ecdhe_ecdsa_aes_128_sha", false);
user_pref("security.ssl3.ecdhe_rsa_aes_128_sha", false);
/* 1263: disable DHE (Diffie-Hellman Key Exchange)
 * [WARNING] may break obscure sites, but not major sites, which should support ECDH over DHE
 * [1] https://www.eff.org/deeplinks/2015/10/how-to-protect-yourself-from-nsa-attacks-1024-bit-DH ***/
user_pref("security.ssl3.dhe_rsa_aes_128_sha", false);
user_pref("security.ssl3.dhe_rsa_aes_256_sha", false);
/* 1264: disable the remaining non-modern cipher suites as of FF52
 * [NOTE] commented out because it still breaks too many sites ***/
   // user_pref("security.ssl3.rsa_aes_128_sha", false);
   // user_pref("security.ssl3.rsa_aes_256_sha", false);
/* 1265: block rc4 fallback (will be deprecated in 53) ***/
user_pref("security.tls.unrestricted_rc4_fallback", false);
/* UI (User Interface) ***/
/* 1270: display warning (red padlock) for "broken security"
 * [1] https://wiki.mozilla.org/Security:Renegotiation ***/
user_pref("security.ssl.treat_unsafe_negotiation_as_broken", true);
/* 1271: control "Add Security Exception" dialog on SSL warnings
 * 0=do neither 1=pre-populate url 2+pre-populate url + pre-fetch cert (default)
 * [1] https://github.com/pyllyukko/user.js/issues/210 ***/
user_pref("browser.ssl_override_behavior", 1);
/* 1272: display advanced information on Insecure Connection warning pages
 * only works when it's possible to add an exception
 * i.e doesn't work for HSTS (https://subdomain.preloaded-hsts.badssl.com/)
 * [TEST] https://expired.badssl.com/ ***/
user_pref("browser.xul.error_pages.expert_bad_cert", true);
```

**Comment by Thorin-Oakenpants** [comment]
^^ nothing has changed except the order, descriptions (maybe?), and an intro section paragraph - i.e no values or active status changed

**Comment by earthlng** [comment]
* Do we need ENCRYPTION in the title?
* see [1]. This article
```
see [1] (It's quite technical but the first part is easy to understand
and you can stop reading when you reach the second section titled "Enter Bro")
```
* Maybe make the sub-titles standing out more visually? /** instead of /* ?
=> section titles with 3 *, subs with 2 and items with 1 star
* 1203: I think with TLS1.3 and this, it guarantees perfect Forward-Secrecy, maybe worth a comment, idk
* 1211+1212: if 1212 (.require) "leaks information about the sites you visit to the CA (cert authority)" wouldn't 1211 also? I think that sentence makes more sense under 1211. +the follow-up sentence too, I guess
* 1242: `We disable it because formerly blocked mixed-content may load`
-> that's the "benefit" not the risk. How about: `...because of the additional priming requests, may cause ...`
* 1221: we have a test-page for that, right? add a link?
* 1223: `and is used to always use HTTPS for the domains on that list`
=> `and used to always load those domains over HTTPS` -> sexier and avoid the "is used to use"
* 1241: `leave it this way as too many sites break visually` - i'm using it and am happy with the result. remove the "leave it this way" and word it more neutral and objective, fe. `It breaks many sites visually`.
* 1272: `i.e doesn't work for HSTS`
=> i.e doesn't work for HSTS *conflicts* or *misconfiguration* or *discrepancy* - or something like that

**Comment by Thorin-Oakenpants** [comment]
- encryption has always been in the title, i'll change it to ciphers which makes more sense
- (It's quite technical .. done
- sections /*** subsections /** prefs /* .. done. I've scanned the master js and can't see anything - I think the only other section is the cache one coming up
- 1203 - i don't understand. SSleuth shows me I visit sites with PFS already. How exactly does this guarantee that? I think I'll just leave the info as is, since it will be the default at 4 soon in 5 weeks.
- 1211+1212: those are the original descriptions we've always had, Maybe Francois can elaborate. I'll post him a Q after this.
- 1242: is this ok
  * Allowing HSTS Priming may load formerly blocked mixed-content, but it also generates additional priming requests which may cause noticeable delays eg requests time out or are not handled well by servers, and there are possible fingerprinting issues
- 1221: [TEST] https://fiprinca.0x90.eu/poc/  (also just added it to the wiki under Intermediate CA Cache Fingerprinting)
- 1223: roger roger, that's some messed up sentence right there- I swear u wrote that when we talked about it on ghacks, or in fact on here when we did the PK active values compare and flipped it from inactive disable to active enforce :)
- 1241: yup:  * [WARNING] when set to true, this will visually break many sites (March 2017) ***/
- 1272: you wrote that man - see: https://github.com/ghacksuserjs/ghacks-user.js/issues/26#issuecomment-282012261. What do you want it to say then?

**Comment by Thorin-Oakenpants** [comment]
Q for @fmarier Here is what we have in the js (as well as `security.ssl.enable_ocsp_stapling`->true)

```js
/* 1211: query OCSP responder servers to confirm current validity of certificates
 * 0=disable, 1=validate only certificates that specify an OCSP service URL (default)
 * 2=enable and use values in security.OCSP.URL and security.OCSP.signing ***/
user_pref("security.OCSP.enabled", 1);
/* 1212: require certificate revocation check through OCSP protocol
 * This leaks information about the sites you visit to the CA (cert authority)
 * It's a trade-off between security (checking) and privacy (leaking info to the CA) ***/
   // user_pref("security.OCSP.require", true);
```

The question is about the comment re "this leaks information...trade-off". Can you ELI5 the difference between 1211 and 1212 as to WHEN these are triggered. When is a OCSP responder server contacted, when is a revocation check done. Would that sentence apply to both (note: in this JS I would push for the check, 1212 is only inactive due to site breakage) - TIA

**Comment by earthlng** [comment]
* 1203: I remember the talk about TLS1.3 at the CCCcon mentioned that the session ids/ticket still allow for cracking previously stored encrypted sessions. but yeah, we can ignore it I guess.
* 1211+1212: well, based on the names one **enables** the use of it and the other **requires** the use of it, afaik. But if 1211 is disabled I'd assume 1212 is irrelevant. Hence the sentences should be under 1211.
Will take a look at the code and see if that's the case.
* 1242 `is this ok?` I see you want to keep the benefit, makes sense, but I'd still slightly change it:
Allowing HSTS Priming may load formerly blocked mixed-content, but it ~~also~~ **does so by** ~~generates~~ **sending** additional priming requests ...
* 1223: `I swear u wrote that when we talked` - that may well be true but at that point I didn't care about wording it nicely.
* 1272: `you wrote that man` - again, at that point it illustrated what I meant, but re-reading it now, I felt it could use some improving, or it could be misunderstood as meaning it doesn't work when HSTS is enabled or something. `i.e doesn't work for HSTS discrepancies` seems fine imo. I just didn't know if **discrepancies** is the right word to use here, so I listed some other ones to choose from.

**Comment by Thorin-Oakenpants** [comment]
^^ all done. Just waiting for Francois to maybe elaborate re 1211+1212. I most likely added them weeks/months apart way back and I vaguely remember talking to moonchild about it in an email (it was a round the time I setup a palemoon section, that's how long ago it was). So it probably is true for both. Should probably remove the comment - I mean the O in OCSP stands for Online - its inferred there will be a third party

**Comment by Atavic** [comment]
@earthlng There's a section titled: Weak SSL/TLS Ciphers/Protocols/Keys

The page is a complete explanation.

**Comment by earthlng** [comment]
> Should probably remove the comment

No, I'd leave it

**Comment by fmarier** [comment]
> Just waiting for Francois to maybe elaborate re 1211+1212.

I'm certainly not an expert on TLS, but in my experience, OCSP (the non-stapled kind) can be quite annoying when OCSP servers are down or when they are blocked by a captive portal. So I personally just use the defaults there (`enabled = 1` but `required = false`).

**Comment by fmarier** [comment]
> something I wouldn't mind seeing "fixed" directly in the MCB itself, would be to upgrade HTTP to HTTPS for resources on domains that already have a HSTS entry in the cache or the preload list. [...] That way no additional requests would be necessary and MCB could redirect internally to the HTTPS resource.

As I understand it, this is what `use_hsts` does. It uses the *existing* HSTS information to automatically upgrade what would otherwise be blocked, without generating extra requests. `send_hsts_priming` is what sends the extra requests to try and discover new HSTS domains. It sounds like you'd be happy with `use_hsts = true` and `send_hsts_priming = false`?

> I just really don't like the additional "priming" requests.

That's the part I don't understand. I'm not sure why these requests are seen as undesirable.

**Comment by Thorin-Oakenpants** [comment]
https://en.wikipedia.org/wiki/Ocsp -> EIL5

@earthlng sure, `.enabled` is the master switch, `.required` is enforcing it (breakage: see francois' comment). Will move the "leaks/trade-off" comment from 1212 to 1211.

Done, will post commit and close

**Comment by Thorin-Oakenpants** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/commit/b2eccc65f425689b86d1de39767f56d8ceef01f9

**Comment by earthlng** [comment]
@fmarier, thanks for your reply.

> It sounds like you'd be happy with use_hsts = true and send_hsts_priming = false?

Yes absolutely - thanks for pointing it out.
But I originally tested this in an older FF version and it didn't quite work that way.

I've setup a testpage [here](https://earthlng.github.io/testpages/mixedcontent.htm)

In FF51 with `security.mixed_content.use_hsts;true` + `security.mixed_content.send_hsts_priming;false` it never loads the last picture from the domain on the preload-list. FF51 seems to only look in the cache file and ignore the preload list.

The default settings in FF51 were pretty stupid. It sent priming requests but never used them.

> `send_hsts_priming` is what sends the extra requests to try and discover new HSTS domains.

Ok that's new to me. I thought it only sends the priming request when there's already a HSTS entry.
So it's even worse then.

> That's the part I don't understand. I'm not sure why these requests are seen as undesirable.

I mostly don't like it because of what I expect MCB to do. It should block HTTP resources and not try to help me and send countless priming requests to all kinds of domains, regardless if they even support HTTPS.

The new feature is a nice addition but you got me worried when you said you would like to "fix" it in the MCB directly - meaning there wouldn't be a way to disable the priming requests.

Another thing I absolutely don't like is that the Network DevTools never shows the request for the 1st picture when `send_hsts_priming=true`, but that's for another department to fix I guess.
But can you gently kick their asses for me, please ;)
(IMHO the network tab should show EVERY request, always!)

**Comment by earthlng** [comment]
@Thorin-Oakenpants: based on my previous comment we should change the `use_hsts` to true

**Comment by fmarier** [comment]
> Another thing I absolutely don't like is that the Network DevTools never shows the request for the 1st picture when `send_hsts_priming=true`.

That sounds like a bug. I would encourage you to file it :)

**Comment by earthlng** [comment]
> That sounds like a bug. I would encourage you to file it :)

Lol, I'm still waiting for my last bug-report to get some love, but yeah I'll test if it's already fixed in nightly and create a ticket if necessary.

**Comment by Thorin-Oakenpants** [comment]
> @Thorin-Oakenpants: based on my previous comment we should change the use_hsts to true.
It can still be used for fingerprinting but blocking mixed-content does already allow for that anyway.
There's no other downside with using this IMO.

@earthling: so what are you saying: That 1242 (both prefs) are flipped to true, and made active, and the description rewritten from "disable HSTS Priming.." to "enforce HSTS Priming.." ?

**Comment by earthlng** [comment]
No, activate both prefs and set `use_hsts=true`. It's stupid when `send_hsts_priming=true` + `use_hsts=false` because it sends the priming requests but never uses them. `use_hsts` is like the master switch and `send_hsts_priming` another feature on top of that to discover new HSTS supporting sites for otherwise blocked requests.

/* 1242: allow Mixed-Content-Blocker to use the HSTS cache but disable the HSTS Priming requests (FF51+)
 * Allow resources from domains with an existing HSTS cache entry or in the HSTS preload list to be
 * upgraded to HTTPS internally but disable sending out HSTS Priming requests, because those may cause
 * noticeable delays eg requests time out or are not handled well by servers, and there are possible
 * fingerprinting issues
 * [NOTE] if you want to use the priming requests make sure 'use_hsts' is true also
 * [1] https://bugzilla.mozilla.org/show_bug.cgi?id=1246540#c145 ***/
user_pref("security.mixed_content.use_hsts", true);
user_pref("security.mixed_content.send_hsts_priming", false);

or we could split them up as 1242 + 1242b, idk, what do you think?

**Comment by fmarier** [comment]
>  because those may cause noticeable delays eg requests time out or are not handled well by servers

If you witness any of these things, a bug should be filed. The priming requests are meant to be unnoticeable so if they cause any issues, that's clearly a bug in the implementation.

> and there are possible fingerprinting issues

Can you explain who would be doing the fingerprinting and how they would do it? I don't see it.

**Comment by Atavic** [comment]
```
user_pref("security.mixed_content.use_hsts", true);
user_pref("security.mixed_content.send_hsts_priming", false);
```

Seems like an half used feature, either use full hsts options or change both entries to false.

**Comment by crssi** [comment]
@fmarier  see this page [http://www.radicalresearch.co.uk/lab/hstssupercookies/](http://www.radicalresearch.co.uk/lab/hstssupercookies/)

**Comment by earthlng** [comment]
@fmarier, sorry for the delay

> > because those may cause noticeable delays eg requests time out or are not handled well by servers
> 
> If you witness any of these things, a bug should be filed. The priming requests are meant to be unnoticeable so if they cause any issues, that's clearly a bug in the implementation.

You already had to decrease the timeout from I think 10sec to 3sec. Delays are not necessarily an implementation bug. fe. site that use lots of images from different very slow servers. Also if sites do support HTTPS but not HSTS (fe. the ghacks image in my testpage) there's the overhead to establish the secure connection (right?). Your own test-server didn't handle HEAD requests all to well. Firewalls could drop requests to 443 leading to the full 3sec timeout. I could probably go on but I think you see my point. Unless mozilla plans to sponsor faster servers in some of those cases there's really not much more you can do to improve the situation.

> > and there are possible fingerprinting issues
>
> Can you explain who would be doing the fingerprinting and how they would do it? I don't see it.

Users A don't block anything and request all resources (X) from a server
Users B block mixed-content (passive and/or active) and request resources X-Y
Users C also send some HEAD requests.

Add the 24h cache to that and the server could even know if you closed FF in the meantime and/or deleted siteSettings.

Clearly those groups of users are easily discernible wouldn't you say? Isn't that what's considered one or more bits of fingerprints?

This could probably all be done from a single domain with a number of subdomains - so "who would be doing the fingerprinting" - anyone who wants to.

You're directly at the source. You could ask Arthur or whoever would be more knowledgeable about those things. If you do that please let me know what they told you.

**Comment by earthlng** [comment]
@Atavic

> Seems like an half used feature, either use full hsts options or change both entries to false.

We block most or all unsolicited requests with this user.js so our current settings are perfectly fine. You can set both prefs to either true or false in your own user.js
